src/repart/repart.c

```diff
@@ -5827,76 +5823,97 @@ static int add_subvolume_path(const char *path, Set **subvolumes) {
         if (r < 0)
                 return log_error_errno(r, "Failed to stat source file '%s/%s': %m", strempty(arg_copy_source), path);
 
-        r = set_ensure_consume(subvolumes, &inode_hash_ops, TAKE_PTR(st));
+        r = hashmap_ensure_put(subvolumes, &inode_hash_ops, st, UINT_TO_PTR(flags));
         if (r < 0)
                 return log_oom();
+        if (r == 0)
+                log_debug("Found duplicate subvolume path %s for the same partition", path);
+
+        TAKE_PTR(st);
 
         return 0;
 }
 
-static int make_subvolumes_strv(const Partition *p, char ***ret) {
-        _cleanup_strv_free_ char **subvolumes = NULL;
+static int make_subvolumes_hashmap(const Partition *p, Hashmap **ret) {
+        _cleanup_hashmap_free_ Hashmap *hashmap = NULL;
         Subvolume *subvolume;
         int r;
 
         assert(p);
         assert(ret);
 
-        ORDERED_HASHMAP_FOREACH(subvolume, p->subvolumes)
-                if (strv_extend(&subvolumes, subvolume->path) < 0)
+        ORDERED_HASHMAP_FOREACH(subvolume, p->subvolumes) {
+                _cleanup_free_ char *path = NULL;
+
+                path = strdup(subvolume->path);
+                if (!path)
+                        return log_oom();
+
+                r = hashmap_ensure_put(&hashmap, &string_hash_ops_free, path, UINT_TO_PTR(subvolume->flags));
+                if (r < 0)
```

Comment:
Inserting into a hashmap can fail for more than just OOM reasons, so `log_oom()` here is wrong.

Reply:
Right. Replaced with `return r;